File: Packages/com.battlement.client/Runtime/Protocol/Ids.cs
#nullable enable

using System;

namespace Battlement
{
    /// <summary>Identifies one connection or reconnection session.</summary>
    public readonly struct SessionId : IEquatable<SessionId>
    {
        /// <summary>Creates an identifier from its nonzero UUID value.</summary>
        public SessionId(Guid value) => Value = value;

        /// <summary>Gets the underlying UUID.</summary>
        public Guid Value { get; }

        public bool Equals(SessionId other) => Value.Equals(other.Value);

        public override bool Equals(object? obj) => obj is SessionId other && Equals(other);

        public override int GetHashCode() => Value.GetHashCode();

        public override string ToString() => Value.ToString();

        public static bool operator ==(SessionId left, SessionId right) => left.Equals(right);

        public static bool operator !=(SessionId left, SessionId right) => !left.Equals(right);
    }

    /// <summary>Identifies one client action for session-wide duplicate detection.</summary>
    public readonly struct ActionId : IEquatable<ActionId>
    {
        /// <summary>Creates an identifier from its nonzero UUID value.</summary>
        public ActionId(Guid value) => Value = value;

        /// <summary>Gets the underlying UUID.</summary>
        public Guid Value { get; }

        public bool Equals(ActionId other) => Value.Equals(other.Value);

        public override bool Equals(object? obj) => obj is ActionId other && Equals(other);

        public override int GetHashCode() => Value.GetHashCode();

        public override string ToString() => Value.ToString();

        public static bool operator ==(ActionId left, ActionId right) => left.Equals(right);

        public static bool operator !=(ActionId left, ActionId right) => !left.Equals(right);
    }

    /// <summary>Identifies one ordered batch of commands.</summary>
    public readonly struct BatchId : IEquatable<BatchId>
    {
        /// <summary>Creates an identifier from its nonzero UUID value.</summary>
        public BatchId(Guid value) => Value = value;

        /// <summary>Gets the underlying UUID.</summary>
        public Guid Value { get; }

        public bool Equals(BatchId other) => Value.Equals(other.Value);

        public override bool Equals(object? obj) => obj is BatchId other && Equals(other);

        public override int GetHashCode() => Value.GetHashCode();

        public override string ToString() => Value.ToString();

        public static bool operator ==(BatchId left, BatchId right) => left.Equals(right);

        public static bool operator !=(BatchId left, BatchId right) => !left.Equals(right);
    }

    /// <summary>Identifies one command and any operation started by that command.</summary>
    public readonly struct CommandId : IEquatable<CommandId>
    {
        /// <summary>Creates an identifier from its nonzero UUID value.</summary>
        public CommandId(Guid value) => Value = value;

        /// <summary>Gets the underlying UUID.</summary>
        public Guid Value { get; }

        public bool Equals(CommandId other) => Value.Equals(other.Value);

        public override bool Equals(object? obj) => obj is CommandId other && Equals(other);

        public override int GetHashCode() => Value.GetHashCode();

        public override string ToString() => Value.ToString();

        public static bool operator ==(CommandId left, CommandId right) => left.Equals(right);

        public static bool operator !=(CommandId left, CommandId right) => !left.Equals(right);
    }

    /// <summary>Identifies one game object in a session.</summary>
    public readonly struct ObjectId : IEquatable<ObjectId>
    {
        /// <summary>Creates an identifier from its nonzero UUID value.</summary>
        public ObjectId(Guid value) => Value = value;

        /// <summary>Gets the underlying UUID.</summary>
        public Guid Value { get; }

        public bool Equals(ObjectId other) => Value.Equals(other.Value);

        public override bool Equals(object? obj) => obj is ObjectId other && Equals(other);

        public override int GetHashCode() => Value.GetHashCode();

        public override string ToString() => Value.ToString();

        public static bool operator ==(ObjectId left, ObjectId right) => left.Equals(right);

        public static bool operator !=(ObjectId left, ObjectId right) => !left.Equals(right);
    }

    /// <summary>Identifies one loaded content-scene instance.</summary>
    public readonly struct SceneId : IEquatable<SceneId>
    {
        /// <summary>Creates an identifier from its nonzero UUID value.</summary>
        public SceneId(Guid value) => Value = value;

        /// <summary>Gets the underlying UUID.</summary>
        public Guid Value { get; }

        public bool Equals(SceneId other) => Value.Equals(other.Value);

        public override bool Equals(object? obj) => obj is SceneId other && Equals(other);

        public override int GetHashCode() => Value.GetHashCode();

        public override string ToString() => Value.ToString();

        public static bool operator ==(SceneId left, SceneId right) => left.Equals(right);

        public static bool operator !=(SceneId left, SceneId right) => !left.Equals(right);
    }
}

File: crates/battlement-types/src/ids.rs
//! Strongly typed identifiers shared by protocol domains.

use std::{error::Error, fmt, marker::PhantomData, str::FromStr};

use serde::{Deserialize, Deserializer, Serialize, Serializer, de};
use uuid::Uuid;

/// The reason a string could not be used as a Battlement identifier.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum IdError {
    /// The value was not a UUID.
    InvalidFormat,
    /// Battlement reserves the all-zero UUID and does not accept it as an ID.
    Nil,
}

impl fmt::Display for IdError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidFormat => formatter.write_str("expected a UUID"),
            Self::Nil => formatter.write_str("the all-zero UUID is not a valid Battlement ID"),
        }
    }
}

impl Error for IdError {}

/// A nonzero UUID tagged at the type level with its protocol role.
///
/// Use the role-specific aliases such as [`SessionId`] and [`ObjectId`] in
/// public APIs. The generic representation centralizes parsing, serialization,
/// and UUID conversion without making identifiers for different protocol roles
/// interchangeable.
#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct ProtocolId<K> {
    uuid: Uuid,
    kind: PhantomData<fn() -> K>,
}

impl<K> ProtocolId<K> {
    /// Creates an identifier from a UUID, rejecting the all-zero value.
    pub const fn from_uuid(uuid: Uuid) -> Result<Self, IdError> {
        if uuid.is_nil() {
            Err(IdError::Nil)
        } else {
            Ok(Self::from_non_nil_uuid(uuid))
        }
    }

    /// Generates a random version-4 identifier.
    #[must_use]
    pub fn new_v4() -> Self {
        Self::from_non_nil_uuid(Uuid::new_v4())
    }

    /// Returns the underlying UUID.
    #[must_use]
    pub const fn as_uuid(&self) -> &Uuid {
        &self.uuid
    }

    /// Consumes this value and returns the underlying UUID.
    #[must_use]
    pub const fn into_uuid(self) -> Uuid {
        self.uuid
    }

    const fn from_non_nil_uuid(uuid: Uuid) -> Self {
        Self {
            uuid,
            kind: PhantomData,
        }
    }
}

impl<K> fmt::Display for ProtocolId<K> {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.uuid.hyphenated().fmt(formatter)
    }
}

impl<K> FromStr for ProtocolId<K> {
    type Err = IdError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        Uuid::parse_str(value)
            .map_err(|_| IdError::InvalidFormat)
            .and_then(Self::from_uuid)
    }
}

impl<K> TryFrom<&str> for ProtocolId<K> {
    type Error = IdError;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        value.parse()
    }
}

impl<K> TryFrom<String> for ProtocolId<K> {
    type Error = IdError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        value.parse()
    }
}

impl<K> From<ProtocolId<K>> for Uuid {
    fn from(value: ProtocolId<K>) -> Self {
        value.uuid
    }
}

impl<K> Serialize for ProtocolId<K> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.uuid.serialize(serializer)
    }
}

impl<'de, K> Deserialize<'de> for ProtocolId<K> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Self::from_uuid(Uuid::deserialize(deserializer)?).map_err(de::Error::custom)
    }
}

mod kind {
    #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
    pub struct Session;

    #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
    pub struct Action;

    #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
    pub struct Batch;

    #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
    pub struct Command;

    #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
    pub struct Object;

    #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
    pub struct Scene;
}

/// Identifies one connection or reconnection session.
pub type SessionId = ProtocolId<kind::Session>;
/// Identifies one client action for session-wide duplicate detection.
pub type ActionId = ProtocolId<kind::Action>;
/// Identifies one ordered batch of commands.
pub type BatchId = ProtocolId<kind::Batch>;
/// Identifies one command and any operation started by that command.
pub type CommandId = ProtocolId<kind::Command>;
/// Identifies one game object in a session.
pub type ObjectId = ProtocolId<kind::Object>;
/// Identifies one loaded content-scene instance.
pub type SceneId = ProtocolId<kind::Scene>;

/// Creates a constant [`ObjectId`] from a UUID literal.
///
/// Invalid or nil UUID literals fail during compilation when used in a constant.
#[macro_export]
macro_rules! object_id {
    ($value:literal) => {{
        const UUID: $crate::__private::Uuid = $crate::__private::uuid!($value);
        match $crate::ObjectId::from_uuid(UUID) {
            Ok(id) => id,
            Err(_) => panic!("object ID must not be nil"),
        }
    }};
}

/// Creates a constant [`SceneId`] from a UUID literal.
///
/// Invalid or nil UUID literals fail during compilation when used in a constant.
#[macro_export]
macro_rules! scene_id {
    ($value:literal) => {{
        const UUID: $crate::__private::Uuid = $crate::__private::uuid!($value);
        match $crate::SceneId::from_uuid(UUID) {
            Ok(id) => id,
            Err(_) => panic!("scene ID must not be nil"),
        }
    }};
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn identifiers_accept_uuid_text_and_reject_nil() {
        let valid = "94fa422b-301d-442d-b9a7-10ea54318e78";
        assert_eq!(valid.parse::<SessionId>().unwrap().to_string(), valid);
        assert!(valid.to_uppercase().parse::<SessionId>().is_ok());
        assert_eq!(
            Uuid::nil().to_string().parse::<SessionId>(),
            Err(IdError::Nil)
        );
    }
}
